File: embeddeds/java/src/main/java/io/github/mosser/arduinoml/embedded/java/dsl/_NodeTreeBuilder.java
package io.github.mosser.arduinoml.embedded.java.dsl;

import io.github.mosser.arduinoml.kernel.behavioral.AnalogCondition;
import io.github.mosser.arduinoml.kernel.behavioral.BooleanCondition;
import io.github.mosser.arduinoml.kernel.behavioral.ConditionTree;
import io.github.mosser.arduinoml.kernel.behavioral.DigitalCondition;
import io.github.mosser.arduinoml.kernel.structural.COMPARATOR;
import io.github.mosser.arduinoml.kernel.structural.Constant;
import io.github.mosser.arduinoml.kernel.structural.OPERATOR;
import io.github.mosser.arduinoml.kernel.structural.SIGNAL;

public class _NodeTreeBuilder {
    TransitionBuilder parent;
    ConditionTree local;
    String subTree;

    public _NodeTreeBuilder(TransitionBuilder parent, String subTree) {
        this.parent = parent;
        this.subTree = subTree;
    }

    public void parseConditionString() {
        // The goal of this function is to recursively create the ConditionTree
        // We will first seek for the most atomic and/or operator
        // The "highest level" is the one that is the least nested in parenthesis
        // Meaning it is the operator that is not in a parenthesis (or in the least nested parenthesis)
        // Then we will create another NodeTreeBuilder for the left and one for the part
        // While setting the operator to the operator we found
        // And left child to the new left NodeTreeBuilder's local, same for the right one

        // First we split the string into tokens
        String[] tokens = subTree.split("\\s+");
        // Remove empty tokens
        tokens = java.util.Arrays.stream(tokens).filter(token -> !token.equals("")).toArray(String[]::new);
        int parenthesisLevel = 0;

        // We will iterate over the tokens to find the highest level operator
        // which depends on the least number of parenthesis
        int highestLevelOperatorIndex = -1;
        int highestLevelOperatorParenthesisLevel = Integer.MAX_VALUE;
        for (int i = 0; i < tokens.length; i++) {
            if (tokens[i].equals("(")) {
                parenthesisLevel++;
            } else if (tokens[i].equals(")")) {
                parenthesisLevel--;
            } else if (tokens[i].equals("&&") || tokens[i].equals("||")) {
                if (parenthesisLevel < highestLevelOperatorParenthesisLevel) {
                    highestLevelOperatorIndex = i;
                    highestLevelOperatorParenthesisLevel = parenthesisLevel;
                }
            }
        }

        // Check if no operator was found
        if (highestLevelOperatorIndex == -1) {

            if(tokens[2].equals("HIGH") || tokens[2].equals("LOW")){
                // Meaning we are on a digital condition
                DigitalCondition node = new DigitalCondition();
                node.setSensor(parent.parent.findSensor(tokens[0]));
                node.setValue(tokens[2].equals("HIGH") ? SIGNAL.HIGH : SIGNAL.LOW);

                local = node;
                if(!tokens[1].equals("==")) {
                    // Error from the dev, we should have == if we are in a digital condition
                    //TODO! Throw a custom exception ?
                    System.out.println("Error: Digital condition should have == as comparator");

                }
            }

            else
            {
                // We are in an analog condition
                // Two possible cases, the value either is the name of a constant
                // Or a direct float value
                boolean isConstant = parent.parent.isConstant(tokens[2]);
                Constant value;

                if(isConstant){
                    value = parent.parent.findConstant(tokens[2]);
                }
                else {
                    // The user directly input a float value
                    // We first check if the entered value isnt already present in on of the constants
                    // If yes, we reuse it, if not we create a new one

                    // Is it actually a float ?
                    try {
                        Float.parseFloat(tokens[2]);
                    } catch (NumberFormatException e) {
                        // Error from the dev, either he entered a wrong variable name or a non float value
                        System.err.println("Error: Invalid value in analog condition, " + tokens[2] +
                                " is not a float nor a known constant");
                        System.exit(1);
                    }

                    Constant constantAlreadyPresent = parent.parent.valueAlreadyPresent(Float.parseFloat(tokens[2]));
                    if(constantAlreadyPresent != null){
                        value = constantAlreadyPresent;
                    }
                    else {
                        value = new Constant(_ConditionTreeBuilder.getNewNameForConstant(), Float.parseFloat(tokens[2]));
                        parent.parent.addConstant(value);
                    }
                }

                AnalogCondition node = new AnalogCondition();
                node.setSensor(parent.parent.findSensor(tokens[0]));
                node.setValue(value);

                // Now we identify the comparator
                switch(tokens[1]){
                    case "<":
                        node.setComparator(COMPARATOR.LT);
                        break;
                    case "<=":
                        node.setComparator(COMPARATOR.LEQ);
                        break;
                    case ">":
                        node.setComparator(COMPARATOR.GT);
                        break;
                    case ">=":
                        node.setComparator(COMPARATOR.GEQ);
                        break;
                    case "==":
                        node.setComparator(COMPARATOR.EQ);
                        break;
                    case "!=":
                        node.setComparator(COMPARATOR.NEQ);
                        break;
                    default:
                        // Error from the dev, we should have a valid comparator
                        //TODO! Throw a custom exception ?
                        System.out.println("Error: Invalid comparator in analog condition");
                }
                local = node;
            }

             return;
        }

        // Now we will build the left subString and the right subString
        String leftSubString = "";
        for (int i = 0; i < highestLevelOperatorIndex; i++) {
            leftSubString += tokens[i] + " ";
        }

        String rightSubString = "";
        for (int i = highestLevelOperatorIndex + 1; i < tokens.length; i++) {
            rightSubString += tokens[i] + " ";
        }

        // Remove outer parentheses from the left and right substrings
        leftSubString = removeOuterParentheses(leftSubString.trim());
        rightSubString = removeOuterParentheses(rightSubString.trim());

        // Now we will create the left and right NodeTreeBuilders
        _NodeTreeBuilder leftNodeTreeBuilder = new _NodeTreeBuilder(parent, leftSubString);
        _NodeTreeBuilder rightNodeTreeBuilder = new _NodeTreeBuilder(parent, rightSubString);

        // Now we will create the local ConditionTree
        local = new BooleanCondition();
        OPERATOR operator = tokens[highestLevelOperatorIndex].equals("&&") ? OPERATOR.AND : OPERATOR.OR;

        // We now delegate the parsing to the left and right NodeTreeBuilders
        leftNodeTreeBuilder.parseConditionString();
        rightNodeTreeBuilder.parseConditionString();

        ((BooleanCondition) local).setOperator(operator);
        ((BooleanCondition) local).setLeftTree(leftNodeTreeBuilder.local);
        ((BooleanCondition) local).setRightTree(rightNodeTreeBuilder.local);
    }

    private String removeOuterParentheses(String str) {
        while (str.startsWith("(") && str.endsWith(")")) {
            str = str.substring(1, str.length() - 1).trim();
        }
        return str;
    }

}

File: embeddeds/python/pyArduinoML/methodchaining/NodeTreeBuilder.py
from pyArduinoML.model.BooleanCondition import BooleanCondition
from pyArduinoML.model.DigitalCondition import DigitalCondition
from pyArduinoML.model.OPERATOR import OPERATOR 
from pyArduinoML.model.SIGNAL import Signal
from pyArduinoML.methodchaining.ConditionTreeBuilder import ConditionTreeBuilder
from pyArduinoML.model.Constant import Constant
from pyArduinoML.model.AnalogCondition import AnalogCondition
from pyArduinoML.model.COMPARATOR import Comparator

class NodeTreeBuilder:
    """
    Builder for recursively creating a ConditionTree from a condition string.
    """

    def __init__(self, parent, sub_tree):
        """
        Initialize the NodeTreeBuilder.

        :param parent: TransitionBuilder, the parent transition builder.
        :param sub_tree: str, the condition string to parse.
        """
        self.parent = parent # Reference to the TransitionBuilder
        self.sub_tree = sub_tree # The condition string to parse
        self.local = None  # The ConditionTree built from this NodeTreeBuilder

    def parse_condition_string(self):
        """
        Recursively parse the condition string and construct the ConditionTree.
        """
        # Split the condition string into tokens
        tokens = self.sub_tree.split()
        tokens = [token for token in tokens if token]  # Remove empty tokens
        parenthesis_level = 0

        # Find the highest-level logical operator (AND/OR)
        highest_level_operator_index = -1
        highest_level_operator_parenthesis_level = float('inf')

        for i, token in enumerate(tokens):
            if token == "(":
                parenthesis_level += 1
            elif token == ")":
                parenthesis_level -= 1
            elif token in ["&&", "||"]:
                if parenthesis_level < highest_level_operator_parenthesis_level:
                    highest_level_operator_index = i
                    highest_level_operator_parenthesis_level = parenthesis_level

        # Check if no operator was found (terminal case)
        if highest_level_operator_index == -1:
            self.terminal_case(tokens)
            return

        # Split into left and right sub-strings
        left_sub_string = " ".join(tokens[:highest_level_operator_index])
        right_sub_string = " ".join(tokens[highest_level_operator_index + 1:])

        # Remove outer parentheses
        left_sub_string = self.remove_outer_parentheses(left_sub_string.strip())
        right_sub_string = self.remove_outer_parentheses(right_sub_string.strip())

        # Recursively create left and right NodeTreeBuilders
        left_node_tree_builder = NodeTreeBuilder(self.parent, left_sub_string)
        right_node_tree_builder = NodeTreeBuilder(self.parent, right_sub_string)

        # Parse the left and right sub-trees
        left_node_tree_builder.parse_condition_string()
        right_node_tree_builder.parse_condition_string()

        # Create a BooleanCondition for this node
        operator = OPERATOR.AND if tokens[highest_level_operator_index] == "&&" else OPERATOR.OR
        self.local = BooleanCondition(
            left_tree=left_node_tree_builder.local,
            right_tree=right_node_tree_builder.local,
            operator=operator
        )


    def terminal_case(self, tokens):
        """
        Handle the terminal case (leaf nodes of the ConditionTree).

        :param tokens: List[str], the tokens representing the terminal condition.
        """
        if tokens[2] in ["HIGH", "LOW"]:
            # DigitalCondition
            node = DigitalCondition()
            node.sensor = self.parent.parent.find_sensor(tokens[0])  # Use attribute assignment
            node.value = Signal.HIGH if tokens[2] == "HIGH" else Signal.LOW  # Use attribute assignment

            self.local = node
            if tokens[1] != "==":
                raise ValueError(f"Digital condition should have '==' as comparator, not: {tokens[1]}")
        else:
            # AnalogCondition
            is_constant = self.parent.parent.is_constant(tokens[2])
            if is_constant:
                value = self.parent.parent.find_constant(tokens[2])
            else:
                try:
                    float_value = float(tokens[2])
                except ValueError:
                    raise ValueError(f"Invalid value in analog condition: {tokens[2]} is not a float nor a known constant")

                value = self.parent.parent.value_already_present(float_value)
                if value is None:
                    value = Constant(ConditionTreeBuilder.get_new_name_for_constant(), float_value)
                    self.parent.parent.add_constant(value)

            node = AnalogCondition()
            node.sensor = self.parent.parent.find_sensor(tokens[0])  # Use attribute assignment
            node.value = value  # Use attribute assignment

            comparator = self.translate_comparator(tokens[1])
            if comparator is None:
                raise ValueError(f"Invalid comparator in analog condition: {tokens[1]}")
            node.comparator = comparator  # Use attribute assignment

            self.local = node


    def translate_comparator(self, token):
        """
        Translate a comparator string into a COMPARATOR enum.

        :param token: str, the comparator token.
        :return: COMPARATOR enum or None if invalid.
        """
        return {
            "<": Comparator.LT,
            "<=": Comparator.LEQ,
            ">": Comparator.GT,
            ">=": Comparator.GEQ,
            "==": Comparator.EQ,
            "!=": Comparator.NEQ
        }.get(token, None)

    def remove_outer_parentheses(self, string):
        """
        Remove outer parentheses from a string.

        :param string: str, the input string.
        :return: str, the string without outer parentheses.
        """
        while string.startswith("(") and string.endswith(")"):
            string = string[1:-1].strip()
        return string
